src/ffi.lisp

```diff
@@ -580,86 +582,71 @@ session-resume requests) would normally be copied into the local cache before pr
 
 #+clozure-common-lisp
-(defun output-wait (stream fd deadline)
-  (unless deadline
-    (setf deadline (stream-deadline (ssl-stream-socket stream))))
-  (let* ((timeout
-    (if deadline
-        (milliseconds-until-deadline deadline stream)
-        nil)))
-    (multiple-value-bind (win timedout error)
-  (ccl::process-output-wait fd timeout)
-      (unless win
-  (if timedout
-      (error 'ccl::communication-deadline-expired :stream stream)
-      (ccl::stream-io-error stream (- error) "write"))))))
+(defun io-wait (stream fd deadline direction)
+  (let ((socket (ssl-stream-socket stream)))
+    (unless deadline
+      (setf deadline (stream-deadline socket)))
+    (let ((timeout          (ecase direction
+                              (:input  (stream-input-timeout  socket))
+                              (:output (stream-output-timeout socket))))
+          (deadline-timeout (when deadline
+                              (deadline->timeout deadline)))
+          (timeout-error    (case direction
+                              (:input  'ccl::input-timeout)
+                              (:output 'ccl::output-timeout))))
+      (when (and deadline-timeout (minusp deadline-timeout))
+        (error 'ccl::communication-deadline-expired :stream stream))
+      (when (and timeout (minusp timeout))
+        (error timeout-error :stream stream))
+      (when (or (not timeout)
+                (and deadline-timeout
+                     (< deadline-timeout timeout)))
+        (setf timeout       deadline-timeout
+              timeout-error 'ccl::communication-deadline-expired))
+      (when timeout
+        (setf timeout (round (* 1000 timeout))))
+      (multiple-value-bind (win timedout error)
+          (case direction
+            (:input  (ccl::process-input-wait  fd timeout))
+            (:output (ccl::process-output-wait fd timeout)))
+        (unless win
+          (if timedout
+            (error timeout-error :stream stream)
+            (ccl::stream-io-error stream (- error) "read")))))))
 
 #+sbcl
-(defun output-wait (stream fd deadline)
-  (declare (ignore stream))
-  (let ((timeout
-   ;; *deadline* is handled by wait-until-fd-usable automatically,
-   ;; but we need to turn a user-specified deadline into a timeout
-   (when deadline
-     (/ (- deadline (get-internal-real-time))
-        internal-time-units-per-second))))
-    (sb-sys:wait-until-fd-usable fd :output timeout)))
+(defun io-wait (stream fd deadline direction)
+  (let ((timeout (let ((socket (ssl-stream-socket stream)))
+                   (ecase direction
+                     (:input  (stream-input-timeout  socket))
+                     (:output (stream-output-timeout socket)))))
+        (deadline-timeout (when deadline
+                            (coerce (deadline->timeout deadline)
+                                    'single-float))))
+    (when (and deadline (minusp deadline-timeout))
+      (error 'sb-sys:deadline-timeout :seconds deadline-timeout))
```

Comment:
It can be done, although we have to decide whether we want to store the new deadline in the stream object. If you don't want to support the deadlines coming from SSL-STREAM, this branch of code can be deleted.